include/ofi_recvwin.h

```diff
@@ -85,6 +85,18 @@ ofi_recvwin_queue_msg(struct name *recvq, entrytype * msg, uint64_t id)	\
 	recvq->pending->buf[write_idx] = *msg;				\
 	ofi_cirque_commit(recvq->pending);				\
 	return 0;							\
+}									\
+				                                        \
+static inline entrytype *						\
+ofi_recvwin_get_msg(struct name *recvq, uint64_t id)	   		\
+{		                                           		\
+	int read_idx;							\
+									\
+	assert(ofi_recvwin_is_allowed(recvq, id));			\
+	read_idx = (ofi_cirque_rindex(recvq->pending)			\
+		    + (id - recvq->exp_msg_id))				\
+		    & recvq->pending->size_mask;			\
+	return &recvq->pending->buf[read_idx];				\
```

Comment:
This returns the address of an entry in the array.  The comment in the commit message about returning NULL seems wrong.

Reply:
The array is initialized using calloc(), thus if a message was not set, it will return NULL. I will update the commit message to make it more clear.

Reply:
My mistake, it will return a pointer that pointing to NULL. I will update the comment

Reply:
This is a circular buffer.  As messages are removed, are the entries nulled out, or is the only time null will be returned is the first time through the buffer?

Reply:
Btw, my concern isn't so much about the commit message as callers assuming that (a pointer to) null will be returned if the id is invalid.

Reply:
ofi_recvwin_get_next_msg() is used to remove message, it will null the entry. It will also increase both cirque->rcnt and recvq->exp_msg_id by one, therefore this function will still be able to get correct pointer.  Caller can assume a pointer to NULL will be returned if the id is invalid.

Reply:
Perfect - thanks